crypto/encode_test.go

```diff
@@ -0,0 +1,121 @@
+package crypto
+
+import (
+	"os"
+	"testing"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	tcrypto "github.com/tendermint/tendermint/crypto"
+)
+
+type byter interface {
+	Bytes() []byte
+}
+
+func checkAminoBinary(t *testing.T, src byter, dst interface{}, size int) {
+	// Marshal to binary bytes.
+	bz, err := cdc.MarshalBinaryBare(src)
+	require.Nil(t, err, "%+v", err)
+	// Make sure this is compatible with current (Bytes()) encoding.
+	assert.Equal(t, src.Bytes(), bz, "Amino binary vs Bytes() mismatch")
+	// Make sure we have the expected length.
+	if size != -1 {
+		assert.Equal(t, size, len(bz), "Amino binary size mismatch")
+	}
+	// Unmarshal.
+	err = cdc.UnmarshalBinaryBare(bz, dst)
+	require.Nil(t, err, "%+v", err)
+}
+
+func checkAminoJSON(t *testing.T, src interface{}, dst interface{}, isNil bool) {
+	// Marshal to JSON bytes.
+	js, err := cdc.MarshalJSON(src)
+	require.Nil(t, err, "%+v", err)
+	if isNil {
+		assert.Equal(t, string(js), `null`)
+	} else {
+		assert.Contains(t, string(js), `"type":`)
+		assert.Contains(t, string(js), `"value":`)
+	}
+	// Unmarshal.
+	err = cdc.UnmarshalJSON(js, dst)
+	require.Nil(t, err, "%+v", err)
+}
+
+func ExamplePrintRegisteredTypes() {
+	cdc.PrintTypes(os.Stdout)
+	// Output: | Type | Name | Prefix | Length | Notes |
+	//| ---- | ---- | ------ | ----- | ------ |
+	//| PrivKeyLedgerSecp256k1 | tendermint/PrivKeyLedgerSecp256k1 | 0x10CAB393 | variable |  |
+	//| PubKeyEd25519 | tendermint/PubKeyEd25519 | 0x1624DE64 | 0x20 |  |
+	//| PubKeySecp256k1 | tendermint/PubKeySecp256k1 | 0xEB5AE987 | 0x21 |  |
+	//| PrivKeyEd25519 | tendermint/PrivKeyEd25519 | 0xA3288910 | 0x40 |  |
+	//| PrivKeySecp256k1 | tendermint/PrivKeySecp256k1 | 0xE1B0F79B | 0x20 |  |
+	//| SignatureEd25519 | tendermint/SignatureEd25519 | 0x2031EA53 | 0x40 |  |
+	//| SignatureSecp256k1 | tendermint/SignatureSecp256k1 | 0x7FC4A495 | variable |  |
+}
+
+func TestKeyEncodings(t *testing.T) {
+	cases := []struct {
+		privKey           tcrypto.PrivKey
+		privSize, pubSize int // binary sizes
+	}{
+		{
+			privKey:  tcrypto.GenPrivKeyEd25519(),
+			privSize: 69,
```

Comment:
Can these numbers be abstracted out into constants with numbers for how they were derived? Right now they seem like magic numbers. (I imagine the answer is 64 bytes for ed25519 privkey + 5 bytes of amino overhead, but would rather this written down then me inferring)

Doesn't block this PR, can easily be done later. I'll make an issue for this

Reply:
Made an issue in #1410